### PR TITLE
MouseEvent and MouseListener

### PR DESCRIPTION
**Pull request by darkdukey**
Add MouseEvent and Mouse Listener to handle mouse input in desktop systems.

### CODE REVIEW

**Comment by dumganhar** [comment]
Hi @darkdukey ,
This PR looks great, thanks in advance. :)

File: cocos/2d/CCEventListenerMouse.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    static EventListenerMouse* create();
+
+    /// Overrides
+    virtual EventListenerMouse* clone() override;
+    virtual bool checkAvailable() override;
+
+    std::function<void(Event* event)> onMouseDown;
+    std::function<void(Event* event)> onMouseUp;
+    std::function<void(Event* event)> onMouseMove;
+    std::function<void(Event* event)> onMouseScroll;
```
Comment (dumganhar) [inline cocos/2d/CCEventListenerMouse.h:48]: What do you think of defining the callback to `std::function<void(EventMouse* event)>` to avoid cast?

Comment (darkdukey) [inline cocos/2d/CCEventListenerMouse.h:48]: Sounds good,

Can IDE (xcode, VS) recognize void(EventMouse\* event) and show user hints ? 

-Nite

Comment (dumganhar) [inline cocos/2d/CCEventListenerMouse.h:48]: No, but if you use a different type, compiler will trigger an error. So I think it isn't relevant to change it `std::function<void(EventMouse* event)>`. What do you think?

**Comment by dumganhar** [comment]
Thank you.

**Pull request merged**